### PR TITLE
Fix FrameworkProcessor for SageMaker Pipelines

### PR DESCRIPTION
**Pull request by athewsey**
**Issue #, if available:** #2656, #2763

**Description of changes:**

A slimmed-down alternative to #2633, reducing unnecessary change to `ScriptProcessor` to try and speed up acceptance so we can defer longer-term refactoring proposals to later.

- Fix `FrameworkProcessor`-based processors' interactions with SageMaker Pipelines `ProcessingStep` (which currently don't work together at all per the linked issue)
- Enable `source_dir` via `ProcessingStep`, so Pipelines users can take full advantage of the FrameworkProcessor functionality.
- Fix pass-through of KMS configuration from `FrameworkProcessor` to the underlying Framework Estimator.
- Re-align FrameworkProcessor from `/bin/bash` to `/bin/sh`: Since some optimized, container-oriented Linux distributions (e.g. Alpine) may not include bash.
- Update the dev guide docs, which were never updated to include the FrameworkProcessor classes in API tables of contents or mention available frameworks in the overview.
- (2022-01-27 update) Add `code_location` parameter to ScriptProcessor (including SKLearnProcessor) as well, to enable users of these processors to avoid using the SageMaker default bucket.

**Testing done:**

- Added unit test to prove the fix (that `_normalize_args` as called by Pipelines should fully prepare a `FrameworkProcessor`)
- Adjusted unit tests to further exercise `FrameworkProcessor` on other frameworks
- Adjusted pipelines integration test for FrameworkProcessor to actually check the job is able to run and succeed.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [X] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [X] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [X] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ahsan-z-khan** [comment]
Is this one needs to merged before #2564?

File: src/sagemaker/processing.py [hunk 1]
```diff
@@ -1265,9 +1266,8 @@ class FeatureStoreOutput(ApiObject):
 class FrameworkProcessor(ScriptProcessor):
     """Handles Amazon SageMaker processing tasks for jobs using a machine learning framework."""
 
-    framework_entrypoint_command = ["/bin/bash"]
+    framework_entrypoint_command = ["/bin/sh"]
```
[the remark this answers is not in the document]
Comment (athewsey) [reply src/sagemaker/processing.py:1269]: This change will somewhat restrict the available syntax for the framework script, but that's private logic (`_generate_framework_script`). It won't change anything about the execution environment which is still `python` per line 1346 (should this actually be changed to `python3` maybe?). The main effect it will have on users is compatibility with different container images by standardising on POSIX-compliant `sh` rather than the more advanced `bash` shell which some OSs don't offer out-of-the-box (e.g. [Alpine Linux](https://wiki.alpinelinux.org/wiki/How_to_get_regular_stuff_working) or others mentioned [here](https://stackoverflow.com/a/5725402)). While I'm aware of multiple distros that offer some `sh` implementation but not `bash` in particular, I've not found any that go the other way around.

From my investigations, Ubuntu (16+) is used as a base by all of the [DLCs](https://github.com/aws/deep-learning-containers), the [Scikit-Learn framework container](https://github.com/aws/sagemaker-scikit-learn-container), the [XGBoost framework container](https://github.com/aws/sagemaker-xgboost-container), and even the [Chainer container](https://github.com/aws/sagemaker-chainer-container). As per [here](https://wiki.ubuntu.com/DashAsBinSh), Ubuntu has supported both `/bin/sh` and `/bin/bash` since at least v6.

The [Spark container](https://github.com/aws/sagemaker-spark-container) (which isn't actually FrameworkProcessor-based) seems to use Amazon Linux 2, which also offers `/bin/sh`.

Overall from what I can tell this is not breaking and from a compatibility standpoint should be a step forward rather than back?

File: src/sagemaker/processing.py [hunk 2]
```diff
@@ -1589,41 +1540,161 @@ def run(  # type: ignore[override]
...
 
-        estimator = self._upload_payload(
+        inputs_with_code = self._include_code_in_inputs(
+            inputs,
             code,
-            source_dir,
-            dependencies,
-            git_config,
-            job_name,
+            source_dir=source_dir,
```
[the remark this answers is not in the document]
Comment (athewsey) [reply src/sagemaker/processing.py:1619]: Shouldn't be *necessary* in the sense that the order here matches the function declaration so it'd still work.

I did it for these because:
- They're defined as keyword args and the order didn't seem obvious as a user
- The fact that FrameworkProcessor inserts `source_dir` into the arg order defined by the parent, rather than only appending arguments in the child class, meant that being a bit more explicit could maybe help avoid confusion.

They can go back to positional if needed I'd say

**Review by jeniyat — DISMISSED** [review]
lgtm

**Comment by athewsey** [comment]
`-slow-tests` timed out. `-pr` looks like an unrelated failure that `predictor.delete_model()` in `test_kmeans` should probably be in a try/except in case of timeout?

**Comment by athewsey** [comment]
^ slow-tests timed out again

**Review by navaj0 — APPROVED** [review]

**Comment by shreyapandit** [comment]
> The only notable conflict in this merge was that somebody had tried to introduce `kms_key` argument explicitly through `ProcessingStep`, whereas this PR was already proposing to allow `**kwargs` to pass through open-ended arguments to the Processor (which would have included support for `kms_key`).
> 
> Therefore to resolve the conflict I **removed** the new explicit `kms_key` argument to prevent inconsistency and bloat in the interface. This means users would still be able to pass `kms_key` in, but no longer be able to access the property `step.kms_key` (should have `step.processing_kwargs["kms_key"]` if they really needed it). If this is too much change (reviewers?) then I guess we'll have to go back to having both an explicit `kms_key` argument and kwargs support for all the others 🤢
> 
> I saw a weird error locally in `tests/integ/test_workflow_with_clarify.py` (execution.describe() only found one step instead of 2?) but hoping it's just an environment issue as don't think this merge should have affected it... Will see what CI finds.

I agree with keeping the interface clean and allowing this to be passed through kwargs if everything else also is atm, lets check in with @jayatalr to ensure nothing breaks on the pipelines end of things

**Comment by shreyapandit** [comment]
I see a bunch of other conflicts when I try to rebase with the latest dev branch, @athewsey can you please look into those? The main files with issues are steps.py, workflow.py and related tests. I gave a shot at resolving but you are the best judge of the changes that went in there. Once the file diff is a little cleaner, happy to re-review this again

File: src/sagemaker/fw_utils.py [hunk 3]
```diff
@@ -74,6 +74,8 @@
         "2.6",
         "2.6.0",
         "2.6.2",
+        "2.7",
+        "2.7.0",
```
Comment (shreyapandit) [inline src/sagemaker/fw_utils.py:78]: Has this changed since and we need to update it?

File: src/sagemaker/image_uri_config/tensorflow.json [hunk 4]
```diff
@@ -2629,6 +2661,39 @@
                     "us-west-2": "763104351884"
                 },
                 "repository": "tensorflow-training"
+            },
+            "2.7.0": {
```
Comment (shreyapandit) [inline src/sagemaker/image_uri_config/tensorflow.json:2665]: DLC is yet to release these containers. Can we remove this and introduce this later when DLC officially makes them available?

File: src/sagemaker/image_uri_config/tensorflow.json [hunk 5]
```diff
@@ -1343,6 +1344,36 @@
                     "us-west-2": "763104351884"
                 },
                 "repository": "tensorflow-inference"
+            },
+            "2.7.0": {
+                "registries": {
+                    "af-south-1": "626614931356",
+                    "ap-east-1": "871362719292",
+                    "ap-northeast-1": "763104351884",
```
Comment (shreyapandit) [inline src/sagemaker/image_uri_config/tensorflow.json:1352]: DLC is yet to release these containers. Can we remove this and introduce this later when DLC officially makes them available?

File: src/sagemaker/processing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shreyapandit) [reply src/sagemaker/processing.py:1284]: Thanks, this makes sense

**Comment by mufaddal-rohawala** [comment]
@[athewsey](https://github.com/athewsey) can you help resolve the conflicts in this pr, by rebasing master on this.

**Comment by navinsoni** [comment]
Closing as it's no longer reqd.

**Comment by ChaudharyKaran** [comment]
why is this is not required any more, can we use script processor with code repo's now?

**Comment by gcaria** [comment]
There is a chain of merge requests which point to this MR as the one that should allow for passing additional modules/code to `ScriptProcessor`. Since this MR is now closed, has the issue been addressed somewhere else?
